lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -0,0 +1,412 @@
+                template=PodTemplateSpec(
+                    spec=PodSpec(
+                        containers=[Container(name=container_name, resources=resource_reqs)]
+                    )
+                ),
+            )
+        )
+
+    def _patch(self, _) -> None:
+        """Patch the Kubernetes resources created by Juju to limit cpu or mem."""
+        # Need to ignore invalid input, otherwise the statefulset gives "FailedCreate" and the
+        # charm would be stuck in unknown/lost.
+        try:
+            client = Client()
+            if self._is_patched(client):
+                return
+
+            client.patch(
+                StatefulSet,
+                self._app,
+                self.patched_delta,
+                namespace=self._namespace,
+                patch_type=PatchType.APPLY,
+                field_manager=self.__class__.__name__,
+            )
+
+        except exceptions.ConfigError as e:
+            msg = f"Error creating k8s client: {e}"
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+            return
+
+        except ApiError as e:
+            if e.status.code == 403:
+                msg = f"Kubernetes resources patch failed: `juju trust` this application. {e}"
+            else:
+                msg = f"Kubernetes resources patch failed: {e}"
+
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+
+        except ValueError as e:
+            msg = f"Kubernetes resources patch failed: {e}"
+            logger.error(msg)
+            self.on.patch_failed.emit(message=msg)
+
+        else:
+            logger.info(
+                "Kubernetes resources for app '%s', container '%s' patched successfully: %s",
+                self._app,
+                self._container_name,
+                self.resource_reqs,
+            )
+
+    @classmethod
+    def _conv_res_req(cls, res_req: ResourceRequirements) -> ResourceRequirements:
+        """Convert ResourceRequirements to comparable form.
+
+        - Convert "memory" to GiB.
+        - Convert "cpu" to float.
+
+        When patching the StatefulSet with {"memory": "0.9Gi"}, the actual PodSpec has
+        {"memory": "966367641600m"}; similarly, {"cpu": "0.30000000000000004"} -> {"cpu": "301m"}.
+        So need to parse the strings and convert before comparing.
+        """
+
+        def _conv(dct: ResourceSpecDict) -> Optional[ResourceSpecDict]:
+            """Convert the memory value of a ResourceSpecDict to GiB representation.
+
+            Raises:
+                ValueError, for invalid input.
+            """
+            if not dct:
+                return None
+            copy = dct.copy()
+
+            if memory := copy.get("memory"):
+                if memory.endswith("m"):
+                    # This is milli. Divide by 1000.
+                    value = bitmath.Byte(float(memory[:-1]) / 1000).to_GiB().value
```

Comment:
(Note: it's `bitmath.Byte`, whose `.value` is still a float)
Yes, there could be some unexpected differences because what really needs to be done is round up to 0.001.
I do not plan to merge this before replacing it with lightkube's `parse_quantity` (review pending).